command/agent_test.go

```diff
@@ -620,3 +620,296 @@ listener "tcp" {
+	}
+	defer os.RemoveAll(tmpDirRoot)
+
+	// start test cases here
+	testCases := map[string]struct {
+		templateCount int
+		exitAfterAuth bool
+	}{
+		"zero": {},
+		"zero-with-exit": {
+			exitAfterAuth: true,
+		},
+		"one": {
+			templateCount: 1,
+		},
+		"one_exit": {
+			templateCount: 1,
+			exitAfterAuth: true,
+		},
+		"many": {
+			templateCount: 15,
+		},
+		"many_exit": {
+			templateCount: 15,
+			exitAfterAuth: true,
+		},
+	}
+
+	for tcname, tc := range testCases {
+		t.Run(tcname, func(t *testing.T) {
+			// make some template files
+			var templatePaths []string
+			for i := 0; i < tc.templateCount; i++ {
+				path := makeTempFile(fmt.Sprintf("render_%d", i), templateContents)
+				templatePaths = append(templatePaths, path)
+			}
+
+			// create temp dir for this test run
+			tmpDir, err := ioutil.TempDir(tmpDirRoot, tcname)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// build up the template config to be added to the Agent config.hcl file
+			var templateConfigStrings []string
+			for i, t := range templatePaths {
+				index := fmt.Sprintf("render_%d.json", i)
+				s := fmt.Sprintf(templateConfigString, t, tmpDir, index)
+				templateConfigStrings = append(templateConfigStrings, s)
+			}
+
+			// Create a config file
+			config := `
+vault {
+  address = "%s"
+	tls_skip_verify = true
+}
+
+auto_auth {
+    method "approle" {
+        mount_path = "auth/approle"
+        config = {
+            role_id_file_path = "%s"
+            secret_id_file_path = "%s"
+						remove_secret_id_file_after_reading = false
+        }
+    }
+
+    sink "file" {
+        config = {
+            path = "%s"
+        }
+    }
+}
+
+%s
+
+%s
+`
+
+			// conditionally set the exit_after_auth flag
+			exitAfterAuth := ""
+			if tc.exitAfterAuth {
+				exitAfterAuth = "exit_after_auth = true"
+			}
+
+			// flatten the template configs
+			templateConfig := strings.Join(templateConfigStrings, " ")
+
+			config = fmt.Sprintf(config, serverClient.Address(), roleIDPath, secretIDPath, sinkPath, templateConfig, exitAfterAuth)
+			configPath := makeTempFile("config.hcl", config)
+			defer os.Remove(configPath)
+
+			// Start the agent
+			ui, cmd := testAgentCommand(t, logger)
+			cmd.client = serverClient
+			cmd.startedCh = make(chan struct{})
+
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				code := cmd.Run([]string{"-config", configPath})
+				if code != 0 {
+					t.Errorf("non-zero return code when running agent: %d", code)
+					t.Logf("STDOUT from agent:\n%s", ui.OutputWriter.String())
+					t.Logf("STDERR from agent:\n%s", ui.ErrorWriter.String())
+				}
+				wg.Done()
+			}()
+
+			select {
+			case <-cmd.startedCh:
+			case <-time.After(5 * time.Second):
+				t.Errorf("timeout")
+			}
+
+			// We need to sleep to give Agent time to render the templates. Without this
+			// sleep, the test will attempt to read the temp dir before Agent has had time
+			// to render and will likely fail the test
+			time.Sleep(5 * time.Second)
```

Comment:
I think we could better test on the `exit_after_auth` functionality by only sleeping if that's not set. In the case that exit_after_auth is enabled, we should just let it do the shutdown, and simply wait for the cmd to finish.

Something along these lines:
```
// if using exit_after_auth, then the command will have returned at the
// end and no longer be running. If we are not using exit_after_auth, then
// we need to shut down the command
if !tc.exitAfterAuth {
	// We need to sleep to give Agent time to render the templates. Without this
	// sleep, the test will attempt to read the temp dir before Agent has had time
	// to render and will likely fail the test
	time.Sleep(5 * time.Second)
	cmd.ShutdownCh <- struct{}{}
}
wg.Wait()
```

Reply:
That makes sense; I made the suggested changes in cb9558b29